File: server/src/services/pdf.js
/**
 * Export PDF via Pandoc + WeasyPrint — remplissage A4 une page
 * (port de server/app/services/pdf.py).
 */
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import { countPdfPagesFromFile } from "./pdf-fit.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const SERVER_ROOT = path.resolve(__dirname, "../..");
const STYLE_FILE = path.join(SERVER_ROOT, "style.css");
const ASSETS_DIR = path.join(SERVER_ROOT, "assets");
const PHOTO_FILE = path.join(ASSETS_DIR, "lucas-schrever.jpg");
const PHOTO_PATH = "assets/lucas-schrever.jpg";

function which(cmd) {
  const r = spawnSync(process.platform === "win32" ? "where" : "which", [cmd], {
    encoding: "utf8",
  });
  return r.status === 0;
}

function weasyEnv() {
  const env = { ...process.env };
  if (process.platform === "win32") {
    env.WEASYPRINT_DLL_DIRECTORIES =
      process.env.WEASYPRINT_DLL_DIRECTORIES || "C:\\msys64\\mingw64\\bin";
  }
  return env;
}

function normalizePhotoPaths(markdown) {
  return String(markdown || "")
    .replace(/\]\(lucas-schrever\.jpg\)/g, `](${PHOTO_PATH})`)
    .replace(
      /src=["']lucas-schrever\.jpg["']/g,
      `src="${PHOTO_PATH}"`,
    );
}

function writeOverrideCss(tmpDir, layout) {
  const file = path.join(tmpDir, `cv-scale-${Date.now()}-${Math.random().toString(36).slice(2)}.css`);
  fs.writeFileSync(
    file,
    `:root {\n` +
      `  --cv-scale: ${layout.scale.toFixed(4)};\n` +
      `  --cv-gap-scale: ${layout.gapScale.toFixed(4)};\n` +
      `  --cv-line-height: ${layout.lineHeight.toFixed(3)};\n` +
      `}\n`,
    "utf8",
  );
  return file;
}

function runPandoc(mdPath, pdfPath, overrideCss, env) {
  if (!which("pandoc")) {
    throw new Error("Pandoc introuvable — installe pandoc et ajoute-le au PATH.");
  }
  const args = [
    mdPath,
    "-o",
    pdfPath,
    "--pdf-engine=weasyprint",
    `--css=${pathToFileURL(STYLE_FILE).href}`,
    `--css=${pathToFileURL(overrideCss).href}`,
    `--resource-path=${SERVER_ROOT}`,
  ];
  const r = spawnSync("pandoc", args, {
    cwd: SERVER_ROOT,
    encoding: "utf8",
    env,
  });
  if (r.status !== 0) {
    throw new Error((r.stderr || r.stdout || "Échec Pandoc / WeasyPrint").trim());
  }
}

function countSafe(pdfPath) {
  try {
    return countPdfPagesFromFile(pdfPath);
  } catch {
    return 1;
  }
}

function renderPdf(mdPath, pdfPath, env, layout, tmpDir) {
  const override = writeOverrideCss(tmpDir, layout);
  try {
    runPandoc(mdPath, pdfPath, override, env);
    const pdfBytes = fs.readFileSync(pdfPath);
    const pageCount = countSafe(pdfPath);
    return {
      pdfBytes,
      pageCount,
      scale: layout.scale,
      layout,
      fits: pageCount <= 1,
    };
  } finally {
    try {
      fs.unlinkSync(override);
    } catch {
      /* ignore */
    }
  }
}

function shrinkLayout(layout) {
  return {
    scale: Math.max(0.52, layout.scale - 0.035),
    gapScale: Math.max(0.62, layout.gapScale - 0.045),
    lineHeight: Math.max(1.1, layout.lineHeight - 0.018),
  };
}

function findFittingLayout(mdPath, pdfPath, env, tmpDir) {
  let layout = { scale: 1.0, gapScale: 1.0, lineHeight: 1.28 };
  let last = null;
  for (let i = 0; i < 14; i++) {
    last = renderPdf(mdPath, pdfPath, env, layout, tmpDir);
    if (last.fits) return { layout, best: last };
    layout = shrinkLayout(layout);
  }
  return { layout, best: last };
}

/**
 * Remplit au maximum la page A4 : recherche dichotomique sur l'échelle
 * (et les espacements) depuis un layout qui tient déjà sur 1 page.
 */
function maximizeFill(mdPath, pdfPath, env, base, seedBest, tmpDir) {
  let best = seedBest;
  if (!best || best.pageCount > 1) {
    best = renderPdf(mdPath, pdfPath, env, base, tmpDir);
  }
  if (best.pageCount > 1) return best;

  // Plafond : on cherche le plus grand scale qui reste sur 1 page
  let lo = base;
  let hiScale = Math.min(2.35, base.scale * 1.85);
  let hi = {
    scale: hiScale,
    gapScale: Math.min(1.55, base.gapScale * (hiScale / base.scale)),
    lineHeight: Math.min(1.4, base.lineHeight + 0.1),
  };

  // Si le plafond tient déjà, pousser encore un peu
  let hiTrial = renderPdf(mdPath, pdfPath, env, hi, tmpDir);
  if (hiTrial.fits) {
    best = hiTrial;
    lo = hi;
    hiScale = Math.min(2.6, hi.scale * 1.2);
    hi = {
      scale: hiScale,
      gapScale: Math.min(1.6, lo.gapScale * (hiScale / lo.scale)),
      lineHeight: Math.min(1.42, lo.lineHeight + 0.06),
    };
  }

  for (let i = 0; i < 12; i++) {
    const mid = {
      scale: (lo.scale + hi.scale) / 2,
      gapScale: (lo.gapScale + hi.gapScale) / 2,
      lineHeight: (lo.lineHeight + hi.lineHeight) / 2,
    };
    const trial = renderPdf(mdPath, pdfPath, env, mid, tmpDir);
    if (trial.fits) {
      lo = mid;
      best = trial;
    } else {
      hi = mid;
    }
  }

  return best;
}

function optimizeLayout(mdPath, pdfPath, env, tmpDir) {
  const { layout, best } = findFittingLayout(mdPath, pdfPath, env, tmpDir);
  return maximizeFill(mdPath, pdfPath, env, layout, best, tmpDir);
}

/**
 * @returns {{ pdf: Buffer, filename: string, pageCount: number, scale: number }}
 */
export function exportPdf(markdown, { filename = "cv" } = {}) {
  if (!fs.existsSync(STYLE_FILE)) {
    throw new Error("style.css introuvable à la racine du serveur");
  }
  if (!fs.existsSync(PHOTO_FILE)) {
    throw new Error(`Photo introuvable : ${PHOTO_FILE}`);
  }

  const md = normalizePhotoPaths(markdown);
  const tmp = fs.mkdtempSync(path.join(os.tmpdir(), "cv-pdf-"));
  const mdPath = path.join(tmp, "cv.md");
  const pdfPath = path.join(tmp, "cv.pdf");
  const env = weasyEnv();

  try {
    fs.writeFileSync(mdPath, md, "utf8");
    const result = optimizeLayout(mdPath, pdfPath, env, tmp);
    const safe =
      String(filename || "cv")
        .replace(/[^\w\-]+/g, "-")
        .replace(/-+/g, "-")
        .replace(/^-|-$/g, "") || "cv";
    return {
      pdf: result.pdfBytes,
      filename: `${safe}.pdf`,
      pageCount: result.pageCount,
      scale: result.scale,
    };
  } finally {
    try {
      fs.rmSync(tmp, { recursive: true, force: true });
    } catch {
      /* ignore */
    }
  }
}

File: server/app/services/pdf.py
"""Export PDF via Pandoc + WeasyPrint — fit A4 une page (proportions CSS)."""

from __future__ import annotations

import os
import re
import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path

from fastapi import HTTPException

from app.config import PHOTO_PATH, settings
from app.services.pdf_fit import count_pdf_pages

STYLE_FILE = settings.style_file
ROOT = settings.style_file.parent  # server/


@dataclass
class PdfExportResult:
    pdf_bytes: bytes
    page_count: int
    scale: float
    filename: str


@dataclass(frozen=True)
class _Layout:
    scale: float
    gap_scale: float
    line_height: float


def _normalize_photo_paths(markdown: str) -> str:
    md = str(markdown or "")
    md = md.replace("](lucas-schrever.jpg)", f"]({PHOTO_PATH})")
    md = md.replace("](assets/lucas-schrever.jpg)", f"]({PHOTO_PATH})")
    md = re.sub(
        r'src=["\']lucas-schrever\.jpg["\']',
        f'src="{PHOTO_PATH}"',
        md,
    )
    md = re.sub(
        r'src=["\']assets/lucas-schrever\.jpg["\']',
        f'src="{PHOTO_PATH}"',
        md,
    )
    return md


def _write_override_css(layout: _Layout) -> Path:
    handle = tempfile.NamedTemporaryFile(
        mode="w", suffix=".css", delete=False, encoding="utf-8"
    )
    handle.write(
        ":root {\n"
        f"  --cv-scale: {layout.scale:.4f};\n"
        f"  --cv-gap-scale: {layout.gap_scale:.4f};\n"
        f"  --cv-line-height: {layout.line_height:.3f};\n"
        "}\n"
    )
    handle.close()
    return Path(handle.name)


def _run_pandoc(
    md_path: Path,
    pdf_path: Path,
    override_css: Path,
    *,
    env: dict[str, str],
) -> None:
    subprocess.run(
        [
            "pandoc",
            str(md_path),
            "-o",
            str(pdf_path),
            "--pdf-engine=weasyprint",
            f"--css={STYLE_FILE.resolve().as_uri()}",
            f"--css={override_css.resolve().as_uri()}",
            f"--resource-path={ROOT}",
        ],
        check=True,
        env=env,
        cwd=str(ROOT),
        capture_output=True,
        text=True,
    )


def _count_safe(pdf_bytes: bytes) -> int:
    try:
        return count_pdf_pages(pdf_bytes)
    except Exception:
        return 1


def _render(
    md_path: Path,
    pdf_path: Path,
    *,
    env: dict[str, str],
    layout: _Layout,
) -> PdfExportResult:
    override = _write_override_css(layout)
    try:
        _run_pandoc(md_path, pdf_path, override, env=env)
        pdf_bytes = pdf_path.read_bytes()
        page_count = _count_safe(pdf_bytes)
        return PdfExportResult(
            pdf_bytes=pdf_bytes,
            page_count=page_count,
            scale=layout.scale,
            filename="cv.pdf",
        )
    finally:
        override.unlink(missing_ok=True)


def _shrink(layout: _Layout) -> _Layout:
    return _Layout(
        scale=max(0.52, layout.scale - 0.035),
        gap_scale=max(0.62, layout.gap_scale - 0.045),
        line_height=max(1.1, layout.line_height - 0.018),
    )


def _find_fitting(
    md_path: Path, pdf_path: Path, *, env: dict[str, str]
) -> tuple[_Layout, PdfExportResult]:
    layout = _Layout(scale=1.0, gap_scale=1.0, line_height=1.34)
    last: PdfExportResult | None = None
    for _ in range(14):
        last = _render(md_path, pdf_path, env=env, layout=layout)
        if last.page_count <= 1:
            return layout, last
        layout = _shrink(layout)
    assert last is not None
    return layout, last


def _maximize_fill(
    md_path: Path,
    pdf_path: Path,
    *,
    env: dict[str, str],
    base: _Layout,
    seed: PdfExportResult,
) -> PdfExportResult:
    best = seed
    if best.page_count > 1:
        best = _render(md_path, pdf_path, env=env, layout=base)
    if best.page_count > 1:
        return best

    lo = base
    hi_scale = min(2.35, base.scale * 1.85)
    hi = _Layout(
        scale=hi_scale,
        gap_scale=min(1.55, base.gap_scale * (hi_scale / base.scale)),
        line_height=min(1.4, base.line_height + 0.1),
    )

    hi_trial = _render(md_path, pdf_path, env=env, layout=hi)
    if hi_trial.page_count <= 1:
        best = hi_trial
        lo = hi
        hi_scale = min(2.6, hi.scale * 1.2)
        hi = _Layout(
            scale=hi_scale,
            gap_scale=min(1.6, lo.gap_scale * (hi_scale / lo.scale)),
            line_height=min(1.42, lo.line_height + 0.06),
        )

    for _ in range(12):
        mid = _Layout(
            scale=(lo.scale + hi.scale) / 2,
            gap_scale=(lo.gap_scale + hi.gap_scale) / 2,
            line_height=(lo.line_height + hi.line_height) / 2,
        )
        trial = _render(md_path, pdf_path, env=env, layout=mid)
        if trial.page_count <= 1:
            lo = mid
            best = trial
        else:
            hi = mid

    return best


def _safe_filename(name: str) -> str:
    safe = re.sub(r"[^\w\-]+", "-", str(name or "cv"))
    safe = re.sub(r"-+", "-", safe).strip("-") or "cv"
    return f"{safe}.pdf"


def export_pdf(markdown: str, *, filename: str = "cv") -> PdfExportResult:
    if not STYLE_FILE.exists():
        raise HTTPException(503, "style.css introuvable à la racine du serveur")

    photo = settings.assets_dir / "lucas-schrever.jpg"
    if not photo.exists():
        raise HTTPException(503, f"Photo introuvable : {photo}")

    env = os.environ.copy()
    if os.name == "nt" and Path(settings.weasyprint_dll).is_dir():
        env.setdefault("WEASYPRINT_DLL_DIRECTORIES", settings.weasyprint_dll)

    md = _normalize_photo_paths(markdown)

    with tempfile.TemporaryDirectory(prefix="cv-pdf-") as tmp:
        tmp_path = Path(tmp)
        md_path = tmp_path / "cv.md"
        pdf_path = tmp_path / "cv.pdf"
        md_path.write_text(md, encoding="utf-8")

        try:
            layout, best = _find_fitting(md_path, pdf_path, env=env)
            result = _maximize_fill(
                md_path, pdf_path, env=env, base=layout, seed=best
            )
        except FileNotFoundError as exc:
            raise HTTPException(
                503,
                "Pandoc introuvable — installe Pandoc + WeasyPrint pour l'export PDF.",
            ) from exc
        except subprocess.CalledProcessError as exc:
            detail = (exc.stderr or exc.stdout or str(exc)).strip()
            raise HTTPException(503, f"Échec Pandoc / WeasyPrint : {detail}") from exc

    if result.page_count > 1:
        raise HTTPException(
            503,
            f"Le CV ne tient pas sur une page A4 "
            f"(pages={result.page_count}, scale={result.scale:.3f}). Réduis le contenu.",
        )

    return PdfExportResult(
        pdf_bytes=result.pdf_bytes,
        page_count=result.page_count,
        scale=result.scale,
        filename=_safe_filename(filename),
    )
